src/main/java/com/celements/navigation/cmd/GetNotMappedMenuItemsForParentCommand.java

```diff
@@ -24,42 +24,60 @@
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
 
 import org.apache.commons.logging.Log;
 import org.apache.commons.logging.LogFactory;
+import org.xwiki.context.Execution;
 import org.xwiki.model.reference.DocumentReference;
 
 import com.celements.navigation.IPartNameGetStrategy;
 import com.celements.navigation.TreeNode;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
 import com.xpn.xwiki.XWikiContext;
 import com.xpn.xwiki.XWikiException;
 import com.xpn.xwiki.doc.XWikiDocument;
 import com.xpn.xwiki.objects.BaseObject;
+import com.xpn.xwiki.web.Utils;
 
 public class GetNotMappedMenuItemsForParentCommand {
 
   private static Log LOGGER = LogFactory.getFactory().getInstance(
       GetNotMappedMenuItemsForParentCommand.class);
 
-  private HashMap<String, Map<String, List<TreeNode>>> menuItems;
+  private Map<String, Map<String, List<TreeNode>>> menuItems;
   
   private int queryCount = 0;
 
+  private XWikiContext getContext() {
+    return (XWikiContext) Utils.getComponent(Execution.class).getContext(
+        ).getProperty("xwikicontext");
+  }
+
   public GetNotMappedMenuItemsForParentCommand() {
-    menuItems = new HashMap<String, Map<String, List<TreeNode>>>();
+    menuItems = new ConcurrentHashMap<String, Map<String, List<TreeNode>>>();
```

Comment:
why are you using a ConcurrentHashMap if it is only accessed in synchronized blocks? You may as well be using a HashMap or if possible refactor the code to make use of _putIfAbsent_.